### PR TITLE
Feat: Implement Music Record Token Factory with On-Chain Registry and Deployment Controls

### LINKED ISSUE #13: 🔄 Implement Music Record Token Factory with On-Chain Registry and Deployment Controls
## 📝 Issue Description  
Create a factory contract that can deploy new instances of the Record Share Token contract. This factory should maintain a registry of all deployed tokens and provide functionality to retrieve information about deployed tokens.

## 📌 Requirements  
- Implement a factory contract that can deploy new Record Share Token contracts
- Allow artists or authorized accounts to deploy new tokens via a function like 'deploy_record_token' or whatever name you choose, just ensure its descriptive
- Pass constructor parameters to new tokens (name, symbol, decimals, metadata URI)
- Automatically mint exactly 100 tokens to the deployer (artist)
- Maintain an on-chain registry of all deployed tokens
- Emit a TokenDeployed event when a new token is created
- Include functions to query deployed tokens

## 🔍 Testing Requirements
- Verify the deployment of new Record Share Token contracts with correct parameters
- Test that exactly 100 tokens are minted to the deployer
- Confirm that the factory properly registers newly deployed tokens
- Test permission controls for token deployment (ensure only artists or authorized accounts)
- Verify correct event emissions for token deployment
- Test the get_artist_tokens function returns the correct tokens for a given artist
- Verify that the factory can handle multiple deployments from the same artist
- Test edge cases like deployment with invalid parameters or unauthorized accounts
- Confirm that deployed tokens are fully functional ERC-20 contracts
- Measure and document deployment gas costs

## 🎯 Expected Outcome
- A factory contract that can create new Record Share Tokens, mint the initial 100 tokens, and keep track of all deployments for easy reference.
- Registry system for tracking deployed tokens
- Integration tests showing interaction between factory and deployed tokens
- Documentation of test cases with expected results
- Performance analysis of deployment process.

## 🚀 How to Contribute  

1. **Fork & Clone** the repo.  
2. Create a new branch:  
   ```sh
   git checkout -b [branch-name]


=> Please Provide ETA.

### PR DESCRIPTION
**Pull request by psychemist**
# MusicStrk Pull Request

- [x] Closes #13 
- [x] Added tests
- [x] Run tests
- [x] Run formatting
- [x] Evidence attached
- [x] Commented the code

---

## 📌 Type of Change

- [x] Documentation (updates to README, docs, or comments)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

---

### Artist Role-Based Access Control
- **Permission System:** Only authorized artists or owner can deploy tokens via the factory
- **Role Management:** Owner can grant and revoke artist roles 
- **Security:** Permission checks prevent unauthorized token deployment
- **Flexibility:** Factory owner retains ability to deploy tokens directly

### Token Deployment Options
- **Standardization:** All tokens follow consistent implementation pattern
- **Ownership Transfer:** Factory initially owns tokens then transfers to deployer

### Token Class Hash Management
- **Upgradability:** Owner can update the token implementation class hash
- **Future-Proofing:** Supports implementation upgrades without factory redeployment
- **Access Control:** Updates restricted to owner only
- **Safety Checks:** Prevents setting invalid/zero class hash

---

### Storage Structure
```cairo
#[storage]
struct Storage {
    // ...existing code...
    token_class_hash: ClassHash,
    token_count: u64,
    tokens: Map<u64, ContractAddress>,
    deployed_tokens: Map<ContractAddress, bool>,
    artist_tokens_count: Map<ContractAddress, u64>,
    artist_tokens_items: Map<(ContractAddress, u64), ContractAddress>,
    artist_role: Map<ContractAddress, bool>, // Track artist permissions
}
```

### Artist Role Management

The factory contract tracks authorized artists through a storage mapping with these functions:

```cairo
fn grant_artist_role(ref self: ContractState, artist: ContractAddress);
fn revoke_artist_role(ref self: ContractState, artist: ContractAddress);
fn has_artist_role(self: @ContractState, artist: ContractAddress);
```

### Token Class Hash Management

The contract allows for future token implementation upgrades through class hash management:

```cairo
fn update_token_class_hash(ref self: ContractState, new_class_hash: ClassHash);
fn get_token_class_hash(self: @ContractState) -> ClassHash;
```

### Token Registry Management

The factory maintains a comprehensive tracking system for all deployed tokens:

1. **Global Token Registry**: Sequential index of all tokens
2. **Deployed Token Verification**: Fast lookup for token existence
3. **Artist-Specific Registry**: Tracking which tokens belong to which artists

```cairo
fn get_token_count(self: @ContractState) -> u64;
fn get_token_at_index(self: @ContractState, index: u64) -> ContractAddress;
fn get_tokens_by_artist(self: @ContractState, artist: ContractAddress) -> Array<ContractAddress>;
fn get_all_tokens(self: @ContractState) -> Array<ContractAddress>;
fn is_token_deployed(self: @ContractState, token_address: ContractAddress) -> bool;
```

This design allows efficient querying of:
- All tokens ever deployed
- Tokens deployed by a specific artist
- Token existence verification
- Token metadata and properties

### Deployment Authorization Check

A critical security feature is the permission check during deployment. 
This ensures that only the contract owner or artists with granted permissions can deploy new tokens, preventing unauthorized token creation while maintaining flexibility.

---

### Token Deployment Process

The deployment flow follows these steps:

1. Verify caller has owner or artist role
2. Deploy token with factory as initial owner
3. Initialize token with provided parameters
5. Register token in factory tracking system
6. Transfer token ownership to the token deployer (artist)
7. Emit deployment event with all details

---

### Comprehensive Testing

The test suite includes:

- **Basic Functionality Tests**
  - Successful token deployment
  - Event emission verification
  - Multiple tokens per artist
  - Multiple artists deploying tokens

- **Access Control Tests**
  - Unauthorized deployment attempts
  - Artist role granting and revocation
  - Unauthorized role management

- **Edge Case Tests**
  - Zero address owner rejection
  - Invalid token index handling

- **Performance Analysis**
  - Gas cost comparison between direct and factory deployment
  - Batch deployment efficiency measurement

---

## Benefits

- **Security:** Role-based access control prevents unauthorized token creation
- **Standardization:** All tokens follow the same implementation pattern
- **Efficiency:** Reduces deployment costs for multiple tokens
- **Flexibility:** Supports various token configurations
- **Upgradability:** Token implementation can be updated without redeploying factory
- **Tracking:** Complete visibility over deployed tokens and their creators

---

## 📸 Evidence (A photo is required as evidence)

Screenshot 2025-03-27 at 23 07 58

- scarb build

Screenshot 2025-03-31 at 22 05 17

- scarb test

---

### CODE REVIEW

**Comment by psychemist** [comment]
@CollinsC1O ready for review

File: contract_/src/erc20.cairo [hunk 1]
```diff
@@ -119,6 +122,10 @@ pub mod MusicStrk {
             // Only the owner can initialize the token
             self.ownable.assert_only_owner();
 
+            // Verify caller address is not zero address
+            let caller = get_caller_address();
```
Comment (CollinsC1O) [inline contract_/src/erc20.cairo:126]: ❤️

**Review by CollinsC1O — APPROVED** [review]
